### PR TITLE
FE-291: Refactor query screen on web UI to React

### PR DESCRIPTION
**Pull request by palashkaria**
### Scope & Purpose

- [x] :pizza: New feature

### Checklist

- [x] Tests
  - [x] Manually tested
- [x] :book: CHANGELOG entry made

#### Related Information

- [x] GitHub issue / Jira ticket: https://arangodb.atlassian.net/browse/FE-291

### CODE REVIEW

File: js/apps/system/_admin/aardvark/APP/frontend/js/lib/jquery.csv.min.js
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/frontend/js/lib/jquery.csv.min.js:1]: Looked for all instances of `$.csv` before removing this one

File: js/apps/system/_admin/aardvark/APP/frontend/js/lib/tile.stamen.js
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/frontend/js/lib/tile.stamen.js:1]: switched to a different tiling method (OSM)

File: js/apps/system/_admin/aardvark/APP/react/src/views/graphV2/graphSettings/EdgeStyleSelector.tsx
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/react/src/views/graphV2/graphSettings/EdgeStyleSelector.tsx:1]: in all graphSettings files, extracted the UI component and reused it in the query graph viewer

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryContextProvider.tsx
Comment (palashkaria) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryContextProvider.tsx:1]: this handles editor tab state, including

- editing 
	- query value, bind params
- execution
	- execute/explain/profile
- results
	- removing results
- saving queries
	- save/saveAs/saved list/delete

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/BindVariablesTab.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/BindVariablesTab.tsx:87]: Can we include this element at the interface so it is not TS ignored?

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx:85]:
```suggestion
          <Button colorScheme="gray" onClick={onCloseDebugPackageModal}>
```

**Review by shd8 — COMMENTED** [review]
First review cycle

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:33]:
```suggestion
      setCurrentDisplayType(prevView => (prevView !== "graph" ? "graph" : prevView))
```

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:50]:
```suggestion
      setCurrentDisplayType(prevView => (prevView !== "geo" ? "geo" : prevView);
```

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:58]: Same as above

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:118]:
```suggestion
          if (edge) {
```

**Review by shd8 — COMMENTED** [review]
Second review cycle

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/slowQueries/SlowQueryHistory.tsx
Comment (shd8) [inline js/apps/system/_admin/aardvark/APP/react/src/views/query/slowQueries/SlowQueryHistory.tsx:68]: `TABLE_COLUMNS.map`, `runningQueries.map` and `runningQueries.length === 0` case are very similar to the cases found at `RunningQueries.tsx` file. Maybe it is possible to find an abstraction for them.

**Review by shd8 — COMMENTED** [review]
Third review cycle

File: js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/ControlledJSONEditor.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/ControlledJSONEditor.tsx:71]: That is not possible, hence we needed hack. The underlying JsonEditor npm package doesn't have typed, but it spreads all props, so that makes it work.

File: js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/JSONErrors.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/JSONErrors.tsx:9]: There's no real reason to make this change as it causes no impact. Unless I miss something? Same for all other instances

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryNavigationPrompt.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryNavigationPrompt.tsx:52]: I don't see any particular reason to do so. As above - keeping things nearer helps with less jumping around to gauge what's happening. Especially useful in short components.

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/AQLEditor.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/AQLEditor.tsx:82]: no possible, as mentioned

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/BindVariablesTab.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/BindVariablesTab.tsx:87]: not possible due to the underlying library

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx:85]: I propose we either add some linting for this, or just ignore this. It has actual impact so we can skip this kind of stuff in code reviews in general

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/QueryEditorTopBar.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/QueryEditorTopBar.tsx:44]: very simple logic, best to co-locate

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/SaveAsModal.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/SaveAsModal.tsx:44]: try catch is not required, it's handled inside these functions.

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx:49]: it's not possible as MapContainer from leaflet does not know about the leaflet plugin which needs this; a quirk of integrating external libs :)

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx:66]: do you mean destructure the `type` from geometry?

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGeoView.tsx:112]: are you referring to the `type`? not clear

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGraphView.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/QueryGraphView.tsx:43]: `network?.fit` can be undefined, which means TS will (rightly) complain if I apply this change.

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:33]: I don't see any benefit of using a ternary in this case, infact what it is currently seems cleaner and easier to read

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/queryResults/useDisplayTypes.tsx:50]: same as above

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/runningQueries/RunningQueries.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/runningQueries/RunningQueries.tsx:110]: assuming it's a nit - no impact

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/slowQueries/SlowQueryHistory.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/slowQueries/SlowQueryHistory.tsx]: Good catch! We have a ReactTable abstraction which wasn't available in devel when we started writing this code. Moved it to use that now, in this commit: [`22d4223` (#19380)](https://github.com/arangodb/arangodb/pull/19380/commits/22d4223c24336e04c22607304c662a9b94cf99f4)

File: js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/ControlledJSONEditor.tsx
Comment (palashkaria) [reply js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/ControlledJSONEditor.tsx:119]: this is not possible in most places in this file, as the value coming from props is `any`. If we change this it'll result in breakage in lot of other places. Also - this file already existed before this PR, was just moved to a different location & refactored. Let's keep this change for later, as it's a bigger one

File: js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/JSONErrors.tsx
Comment (shd8) [reply js/apps/system/_admin/aardvark/APP/react/src/components/jsonEditor/JSONErrors.tsx:9]: It was just a suggestion to make it faster to read, it does not have any impact at all. Considering that `!something || !otherThing` is faster to read than `!something || value === othervalues`.

Feel free to resolve this conversation, it's a matter of personal preference.

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryNavigationPrompt.tsx
Comment (shd8) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/QueryNavigationPrompt.tsx:52]: You're right, this particular component is very small, it could be kept as is.

File: js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx
Comment (shd8) [reply js/apps/system/_admin/aardvark/APP/react/src/views/query/editor/DebugPackageModal.tsx]: Agreed, is just a matter of consistency, as most of the function callback props are passed as constants references and not anonymous functions that call other funcs. Feel free to resolve this conversation then.

**Review by shd8 — APPROVED** [review]
All comments have been addressed. After resolving merged conflict it LGTM. Good job.

**Comment by palashkaria** [comment]
Thanks for the review @pluma4345 !

> One minor caveat: because we're using Chakra Tabs instead of routes for the different views, the old URLs for those no longer work (nbd imo) and reloading the page results in the tab being reset (fine by me).

Regarding this: Looks like tabs don't affect the URL currently too - so this doesn't change any behaviour - see screenshot:

image

I'd refrain from breaking any existing behavior, so I double-checked this; I guess these are the tabs you are referring to? Unless I'm missing something else

**Pull request merged**
